mountpoint-s3/src/inode.rs

```diff
@@ -141,19 +141,29 @@ impl Superblock {
                 error!("forget called on inode {ino} already removed from the superblock");
             }
             Some(inode) => {
+                // We have to clone the inode so it will not hold reference to the lock which we want to drop as soon as possible.
+                let inode = inode.clone();
+                drop(inodes);
                 logging::record_name(inode.name());
                 let new_lookup_count = inode.dec_lookup_count(n);
                 if new_lookup_count == 0 {
                     // Safe to remove, kernel no longer has a reference to it.
                     trace!(ino, "removing inode from superblock");
-                    let inode = inodes.remove(&ino).expect("we just retrieved it");
+                    let Some(inode) = self.inner.inodes.write().unwrap().remove(&ino) else {
+                        error!("forget called on inode {ino} already removed from the superblock");
+                        return;
+                    };
 
+                    let inodes = self.inner.inodes.read().unwrap();
```

Comment:
Let's use the scoped lock pattern here.

```suggestion
                    let parent = {
                        let inodes = self.inner.inodes.read().unwrap();
                        inodes.get(&inode.parent()).cloned()
                    }
```